pymc/pytensorf.py

```diff
@@ -78,9 +78,12 @@
 ]
 
 
-def convert_observed_data(data):
+def convert_observed_data(data) -> np.ndarray | Variable:
     """Convert user provided dataset to accepted formats."""
 
+    if isgenerator(data):
+        return floatX(generator(data))
```

Comment:
I think this was on purpose, for something fancy in VI. @ferrine can you confirm whether it's true and whether we still need it?

Reply:
Yes, the refactor only moves the code, but if we can delete it that'd be even better!

Reply:
Doesn't wrapping the generator in floatX consume it immediately?

Reply:
nvm it was also done before

Reply:
User-provided `observed` data comes through this, therefore I'm pretty certain that this generator branch is still relevant for VI.

One next step after this PR could be to split the function into two overloads - one of which applies to `generator` type `data`, and the other to all the rest.

The more important next step should be the introduction of a `dtype` kwarg so we can get #7277 fixed.

Reply:
Just a guess, but that might be for stochastic optimization? If so, that generator could even be infinite.
Why not `(floatX(value) for value in generator)`?

Reply:
If the output can be a generator, the type signature is wrong

Reply:
No, even with a generator, the returned value is a `TensorVariable`:

![grafik](https://github.com/pymc-devs/pymc/assets/5894642/6c8f20a2-47f9-4d60-9bc7-bd965b61cf72)

Reply:
I see some black magic.